app/src/main/java/com/jvoye/tasky/agenda_detail/presentation/AgendaDetailScreenViewModel.kt

```diff
@@ -46,21 +54,114 @@ class AgendaDetailScreenViewModel(
         _state.update { it.copy(
             taskyItem = agendaRepository.getTaskyItem(taskyItemId)
         ) }
+        println("CURRENT STATE: ${_state.value}")
     }
 
-    private fun getNavArgs() {
-        val savedState =  savedStateHandle.get<TaskyNavParcelable>("TaskyNavArgs")
-
+    private fun getTypeAndEditMode(editMode: Boolean, type: TaskyType){
         _state.update { it.copy(
-            taskyItemId = savedState?.taskyItemId,
-            taskyItemType = savedState?.taskyItemType ?: TaskyType.TASK,
-            isEdit = savedState?.isEditMode ?: true
-
+            isEditMode = editMode,
+            taskyItem = it.taskyItem.copy(
+                type = type
+            )
         ) }
+    }
+
+    fun onAction(action: AgendaDetailAction) {
+        when (action) {
+            AgendaDetailAction.OnEditDescriptionClick -> {
+                /*TODO()*/
+            }
+            AgendaDetailAction.OnEditModeClick -> {
+                _state.update { it.copy(
+                    isEditMode = true
+                ) }
+            }
+            AgendaDetailAction.OnEditTitleClick -> {
+                _state.update { it.copy(
+                    isEditMode = true
+                ) }
+            }
+            AgendaDetailAction.OnNotificationTimerClick -> {
+                /*TODO()*/
+            }
+            AgendaDetailAction.OnSaveClick -> {
+                /*TODO()*/
+            }
+            AgendaDetailAction.OnSetDateClick -> {
+                _state.update { it.copy(
+                    isDatePickerDialogVisible = true
+                ) }
+            }
+            AgendaDetailAction.OnSetTimeClick -> {
+                _state.update { it.copy(
+                    isTimePickerDialogVisible = true
+                )}
+            }
+            is AgendaDetailAction.ConfirmDateSelection -> {
 
-        // prints null
-        println("SAVED STATE 2: $savedState")
+                val dateMillis = action.selectedDateMillis
+                val dateInstant = Instant.fromEpochMilliseconds(dateMillis)
+                val newLocalDate = dateInstant.toLocalDateTime(TimeZone.currentSystemDefault()).date
+
+                _state.update { it.copy(
+                    taskyItem = it.taskyItem.copy(
+                        time = newLocalDate.atTime(it.taskyItem.time.hour, it.taskyItem.time.minute)),
+                    isDatePickerDialogVisible = false
+                ) }
+            }
+
+            AgendaDetailAction.OnDismissDatePickerDialog -> {
+                _state.update { it.copy(
+                    isDatePickerDialogVisible = false
+                ) }
+            }
+
+            is AgendaDetailAction.ConfirmTimeSelection -> {
+                val newHour = action.timePickerState.hour
+                val newMinute = action.timePickerState.minute
+                val currentDate = _state.value.taskyItem.time.date
+                val newLocalDateTime = LocalDateTime(
```

Comment:
Suggestion: Instead of rebuilding a whole `LocalDateTime`, you can just send the hour/minute in the action from the state (no need to pass all the state) and use the existing time from state. Then copy it with `withHour` / `withMinute`:

```kotlin
_state.update { it.copy(
    taskyItem = it.taskyItem.copy(
        time = it.taskyItem.time
            .withHour(action.hour)
            .withMinute(action.minute)
    ),
    isTimePickerDialogVisible = false
) }
```